app/models/package_manager/base.rb

```diff
@@ -178,12 +178,18 @@ def self.add_version(db_project, version_hash)
       end
     end
 
-    def self.deprecate_versions(db_project, version_hash)
+    def self.deprecate_versions(db_project, version_hashes)
       case db_project.platform.downcase
-      when "rubygems", "npm" # yanked gems will be omitted from project JSON versions
+      when "rubygems" # yanked gems will be omitted from project JSON versions
         db_project
           .versions
-          .where.not(number: version_hash.pluck(:number))
+          .where.not(number: version_hashes.pluck(:number))
+          .update_all(status: "Removed")
+      when "npm"
+        not_deprecated_version_hashes = version_hashes.select { |v| !v[:is_deprecated] }
+        db_project
+          .versions
+          .where.not(number: not_deprecated_version_hashes.pluck(:number))
```

Comment:
That's fair, I can restructure the query. I did it this way as it was a relatively easy way to cover both the "explicitly deprecated" and the "not present" case, but I can do a little reconfiguring to still cover that.

Reply:
I thought about this more and I think this might be the best approach. If you look at the specs, we want to cover the cases where:
1. We find a version that is deprecated explicitly
2. We find a version that is not in the full list

I think the only other way to handle that than this is to pass in all of the version hashes to filter *out* everything that isn't one of those, and then separately pass in all of the deprecated version hashes to filter *in* everything that matches that.

At that point, I think that this approach is better, but I'm welcome to pushback here if you disagree.